File: libs/monobitmap-manager/src/test/kotlin/mono/bitmap/manager/factory/LineBitmapFactoryTest.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.bitmap.manager.factory

import kotlin.test.Test
import kotlin.test.assertEquals
import mono.graphics.geo.Point
import mono.shape.extra.LineExtra
import mono.shape.extra.manager.predefined.PredefinedStraightStrokeStyle
import mono.shape.extra.style.AnchorChar
import mono.shape.extra.style.StraightStrokeDashPattern

/**
 * A test for [LineBitmapFactory].
 */
class LineBitmapFactoryTest {
    @Test
    fun testToBitmap_simpleHorizontalLine() {
        val points = listOf(
            Point(2, 0),
            Point(6, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals("0───1", bitmap.toString())
    }

    @Test
    fun testToBitmap_simpleHorizontalLine_reversed() {
        val points = listOf(
            Point(6, 0),
            Point(2, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals("1───0", bitmap.toString())
    }

    @Test
    fun testToBitmap_3StraightHorizontalMonotonicPoints() {
        val points = listOf(
            Point(0, 0),
            Point(2, 0),
            Point(4, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals("0───1", bitmap.toString())
    }

    @Test
    fun testToBitmap_3StraightHorizontalMonotonicPoints_reversed() {
        val points = listOf(
            Point(4, 0),
            Point(2, 0),
            Point(0, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals("1───0", bitmap.toString())
    }

    @Test
    fun testToBitmap_3StraightHorizontalNonMonotonicPoints() {
        val points = listOf(
            Point(0, 0),
            Point(4, 0),
            Point(2, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals("0─1──", bitmap.toString())
    }

    @Test
    fun testToBitmap_3StraightHorizontalNonMonotonicPoints_reversed() {
        val points = listOf(
            Point(4, 0),
            Point(0, 0),
            Point(2, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals("──1─0", bitmap.toString())
    }

    @Test
    fun testToBitmap_simpleVerticalLine() {
        val points = listOf(
            Point(0, 2),
            Point(0, 6)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            0
            │
            │
            │
            1
            """.trimIndent(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_simpleVerticalLine_reversed() {
        val points = listOf(
            Point(0, 6),
            Point(0, 2)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            1
            │
            │
            │
            0
            """.trimIndent(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_3StraightVerticalMonotonicPoints() {
        val points = listOf(
            Point(0, 0),
            Point(0, 2),
            Point(0, 4)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            0
            │
            │
            │
            1
            """.trimIndent(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_3StraightVerticalMonotonicPoints_reversed() {
        val points = listOf(
            Point(0, 4),
            Point(0, 2),
            Point(0, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            1
            │
            │
            │
            0
            """.trimIndent(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_3StraightVerticalNonMonotonicPoints() {
        val points = listOf(
            Point(0, 0),
            Point(0, 4),
            Point(0, 2)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            0
            │
            1
            │
            │
            """.trimIndent(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_3StraightVerticalNonMonotonicPoints_reversed() {
        val points = listOf(
            Point(0, 4),
            Point(0, 0),
            Point(0, 2)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            │
            │
            1
            │
            0
            """.trimIndent(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_upperLeft() {
        val points = listOf(
            Point(0, 4),
            Point(4, 4),
            Point(4, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            |    1
            |    │
            |    │
            |    │
            |0───┘
            """.trimMargin(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_upperLeft_reversed() {
        val points = listOf(
            Point(4, 0),
            Point(4, 4),
            Point(0, 4)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            |    0
            |    │
            |    │
            |    │
            |1───┘
            """.trimMargin(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_lowerLeft() {
        val points = listOf(
            Point(0, 0),
            Point(4, 0),
            Point(4, 4)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            |0───┐
            |    │
            |    │
            |    │
            |    1
            """.trimMargin(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_lowerLeft_reversed() {
        val points = listOf(
            Point(4, 4),
            Point(4, 0),
            Point(0, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            |1───┐
            |    │
            |    │
            |    │
            |    0
            """.trimMargin(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_upperRight() {
        val points = listOf(
            Point(4, 4),
            Point(0, 4),
            Point(0, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            |1    
            |│    
            |│    
            |│    
            |└───0
            """.trimMargin(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_upperRight_reversed() {
        val points = listOf(
            Point(0, 0),
            Point(0, 4),
            Point(4, 4)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            |0    
            |│    
            |│    
            |│    
            |└───1
            """.trimMargin(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_lowerRight() {
        val points = listOf(
            Point(4, 0),
            Point(0, 0),
            Point(0, 4)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            |┌───0
            |│    
            |│    
            |│    
            |1    
            """.trimMargin(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_lowerRight_reversed() {
        val points = listOf(
            Point(0, 4),
            Point(0, 0),
            Point(4, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA)
        assertEquals(
            """
            |┌───1
            |│    
            |│    
            |│    
            |0    
            """.trimMargin(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_anchorChar_leftToRight() {
        val anchorCharStart = AnchorChar("id", "name", 'L', 'R', 'T', 'B')
        val anchorCharEnd = AnchorChar("id", "name", 'l', 'r', 't', 'b')
        val lineExtra = LINE_EXTRA.copy(
            userSelectedStartAnchor = anchorCharStart,
            userSelectedEndAnchor = anchorCharEnd
        )
        val points = listOf(
            Point(0, 0),
            Point(4, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, lineExtra)
        assertEquals(
            "L───r",
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_anchorChar_rightToLeft() {
        val anchorCharStart = AnchorChar("id", "name", 'L', 'R', 'T', 'B')
        val anchorCharEnd = AnchorChar("id", "name", 'l', 'r', 't', 'b')
        val lineExtra = LINE_EXTRA.copy(
            userSelectedStartAnchor = anchorCharStart,
            userSelectedEndAnchor = anchorCharEnd
        )
        val points = listOf(
            Point(4, 0),
            Point(0, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, lineExtra)
        assertEquals(
            "l───R",
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_anchorChar_topToBottom() {
        val anchorCharStart = AnchorChar("id", "name", 'L', 'R', 'T', 'B')
        val anchorCharEnd = AnchorChar("id", "name", 'l', 'r', 't', 'b')
        val lineExtra = LINE_EXTRA.copy(
            userSelectedStartAnchor = anchorCharStart,
            userSelectedEndAnchor = anchorCharEnd
        )
        val points = listOf(
            Point(0, 0),
            Point(0, 4)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, lineExtra)
        assertEquals(
            """
            T
            │
            │
            │
            b
            """.trimIndent(),
            bitmap.toString()
        )
    }

    @Test
    fun testToBitmap_anchorChar_bottomToTop() {
        val anchorCharStart = AnchorChar("id", "name", 'L', 'R', 'T', 'B')
        val anchorCharEnd = AnchorChar("id", "name", 'l', 'r', 't', 'b')
        val lineExtra = LINE_EXTRA.copy(
            userSelectedStartAnchor = anchorCharStart,
            userSelectedEndAnchor = anchorCharEnd
        )
        val points = listOf(
            Point(0, 4),
            Point(0, 0)
        )
        val bitmap = LineBitmapFactory.toBitmap(points, lineExtra)
        assertEquals(
            """
            t
            │
            │
            │
            B
            """.trimIndent(),
            bitmap.toString()
        )
    }

    companion object {
        private val LINE_EXTRA = LineExtra(
            isStrokeEnabled = true,
            userSelectedStrokeStyle = PredefinedStraightStrokeStyle.PREDEFINED_STYLES.first(),
            isStartAnchorEnabled = true,
            userSelectedStartAnchor = AnchorChar("id", "name", '0'),
            isEndAnchorEnabled = true,
            userSelectedEndAnchor = AnchorChar("id", "name", '1'),
            dashPattern = StraightStrokeDashPattern.SOLID,
            isRoundedCorner = false
        )
    }
}

File: monosketch-svelte/src/lib/mono/monobitmap/manager/line-bitmap-factory.test.ts
/*
 * Copyright (c) 2024, tuanchauict
 */

import { Point } from "$libs/graphics-geo/point";
import { StringExt } from "$libs/string";
import { PredefinedStraightStrokeStyle } from "$mono/shape/extra/predefined-styles";
import { describe, expect, test } from "vitest";
import { LineBitmapFactory } from './line-bitmap-factory';
import { LineExtra } from '$mono/shape/extra/shape-extra';
import { AnchorChar, StraightStrokeDashPattern } from '$mono/shape/extra/style';
import trimMargin = StringExt.trimMargin;
import trimIndent = StringExt.trimIndent;

const LINE_EXTRA = new LineExtra(
    true,
    PredefinedStraightStrokeStyle.PREDEFINED_STYLES[0],
    true,
    AnchorChar.create({ id: 'id', displayName: 'name', all: '0' }),
    true,
    AnchorChar.create({ id: 'id', displayName: 'name', all: '1' }),
    StraightStrokeDashPattern.SOLID,
    false,
);

describe('LineBitmapFactory.toBitmap', () => {
    describe('simpleHorizontalLine', () => {
        test('simpleHorizontalLine', () => {
            const points = [
                new Point(2, 0),
                new Point(6, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe('0───1');
        });

        test('simpleHorizontalLine_reversed', () => {
            const points = [
                new Point(6, 0),
                new Point(2, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe('1───0');
        });
    });

    describe('3 straight horizontal points', () => {
        test('MonotonicPoints', () => {
            const points = [
                new Point(0, 0),
                new Point(2, 0),
                new Point(4, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe('0───1');
        });

        test('MonotonicPoints_reversed', () => {
            const points = [
                new Point(4, 0),
                new Point(2, 0),
                new Point(0, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe('1───0');
        });

        test('NonMonotonicPoints', () => {
            const points = [
                new Point(0, 0),
                new Point(4, 0),
                new Point(2, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe('0─1──');
        });

        test('NonMonotonicPoints_reversed', () => {
            const points = [
                new Point(4, 0),
                new Point(0, 0),
                new Point(2, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe('──1─0');
        });
    });

    describe('simpleVerticalLine', () => {
        test('simpleVerticalLine', () => {
            const points = [
                new Point(0, 2),
                new Point(0, 6),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimIndent(`
                0
                │
                │
                │
                1
            `),
            );
        });

        test('simpleVerticalLine_reversed', () => {
            const points = [
                new Point(0, 6),
                new Point(0, 2),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimIndent(`
                1
                │
                │
                │
                0
            `),
            );
        });
    });

    describe("3 straight vertical points", () => {
        test('MonotonicPoints', () => {
            const points = [
                new Point(0, 0),
                new Point(0, 2),
                new Point(0, 4),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimIndent(`
                0
                │
                │
                │
                1
        `),
            );
        });

        test('MonotonicPoints_reversed', () => {
            const points = [
                new Point(0, 4),
                new Point(0, 2),
                new Point(0, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimIndent(`
                1
                │
                │
                │
                0
            `),
            );
        });

        test('NonMonotonicPoints', () => {
            const points = [
                new Point(0, 0),
                new Point(0, 4),
                new Point(0, 2),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimIndent(`
                0
                │
                1
                │
                │
            `),
            );
        });

        test('NonMonotonicPoints_reversed', () => {
            const points = [
                new Point(0, 4),
                new Point(0, 0),
                new Point(0, 2),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimIndent(`
                │
                │
                1
                │
                0
            `),
            );
        });
    });

    describe('upperLeft', () => {
        test('upperLeft', () => {
            const points = [
                new Point(0, 4),
                new Point(4, 4),
                new Point(4, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimMargin(`
                |    1
                |    │
                |    │
                |    │
                |0───┘
            `),
            );
        });

        test('upperLeft_reversed', () => {
            const points = [
                new Point(4, 0),
                new Point(4, 4),
                new Point(0, 4),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimMargin(`
                |    0
                |    │
                |    │
                |    │
                |1───┘
            `),
            );
        });
    });

    describe('lowerLeft', () => {
        test('lowerLeft', () => {
            const points = [
                new Point(0, 0),
                new Point(4, 0),
                new Point(4, 4),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimMargin(`
                |0───┐
                |    │
                |    │
                |    │
                |    1
            `),
            );
        });

        test('lowerLeft_reversed', () => {
            const points = [
                new Point(4, 4),
                new Point(4, 0),
                new Point(0, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimMargin(`
                |1───┐
                |    │
                |    │
                |    │
                |    0
            `),
            );
        });
    });

    describe('upperRight', () => {
        test('upperRight', () => {
            const points = [
                new Point(4, 4),
                new Point(0, 4),
                new Point(0, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimMargin(`
                |1    x
                |│    x
                |│    x
                |│    x
                |└───0
            `.replaceAll('x', '')), // x is added to keep the space chars at the end of the line. Stupid lint.
            );
        });

        test('upperRight_reversed', () => {
            const points = [
                new Point(0, 0),
                new Point(0, 4),
                new Point(4, 4),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimMargin(`
                |0    x
                |│    x
                |│    x
                |│    x
                |└───1
            `.replaceAll('x', '')), // x is added to keep the space chars at the end of the line. Stupid lint.
            );
        });
    });

    describe('lowerRight', () => {
        test('lowerRight', () => {
            const points = [
                new Point(4, 0),
                new Point(0, 0),
                new Point(0, 4),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimMargin(`
                |┌───0
                |│    x
                |│    x
                |│    x
                |1    x
            `.replaceAll('x', '')), // x is added to keep the space chars at the end of the line. Stupid lint.
            );
        });

        test('lowerRight_reversed', () => {
            const points = [
                new Point(0, 4),
                new Point(0, 0),
                new Point(4, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, LINE_EXTRA);
            expect(bitmap.toString()).toBe(
                trimMargin(`
                |┌───1
                |│    x
                |│    x
                |│    x
                |0    x
            `.replaceAll('x', '')), // x is added to keep the space chars at the end of the line. Stupid lint.
            );
        });
    });

    describe('anchorChar', () => {
        const anchorCharStart = AnchorChar.create({
            id: 'id',
            displayName: 'name',
            left: 'L',
            right: 'R',
            top: 'T',
            bottom: 'B',
        });

        const anchorCharEnd = AnchorChar.create({
            id: 'id',
            displayName: 'name',
            left: 'l',
            right: 'r',
            top: 't',
            bottom: 'b',
        });

        test('leftToRight', () => {
            const lineExtra = LINE_EXTRA.copy({
                userSelectedStartAnchor: anchorCharStart,
                userSelectedEndAnchor: anchorCharEnd,
            });
            const points = [
                new Point(0, 0),
                new Point(4, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, lineExtra);
            expect(bitmap.toString()).toBe('L───r');
        });

        test('rightToLeft', () => {
            const lineExtra = LINE_EXTRA.copy({
                userSelectedStartAnchor: anchorCharStart,
                userSelectedEndAnchor: anchorCharEnd,
            });
            const points = [
                new Point(4, 0),
                new Point(0, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, lineExtra);
            expect(bitmap.toString()).toBe('l───R');
        });

        test('topToBottom', () => {
            const lineExtra = LINE_EXTRA.copy({
                userSelectedStartAnchor: anchorCharStart,
                userSelectedEndAnchor: anchorCharEnd,
            });
            const points = [
                new Point(0, 0),
                new Point(0, 4),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, lineExtra);
            expect(bitmap.toString()).toBe(
                trimIndent(`
                T
                │
                │
                │
                b
            `),
            );
        });

        test('bottomToTop', () => {
            const lineExtra = LINE_EXTRA.copy({
                userSelectedStartAnchor: anchorCharStart,
                userSelectedEndAnchor: anchorCharEnd,
            });
            const points = [
                new Point(0, 4),
                new Point(0, 0),
            ];
            const bitmap = LineBitmapFactory.toBitmap(points, lineExtra);
            expect(bitmap.toString()).toBe(
                trimIndent(`
                t
                │
                │
                │
                B
            `),
            );
        });
    });
});
